Allow overriding page title and description in FAQs

dmolineus opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #161
| Docs PR or issue | -

<!--
Bugfixes should be based on the 4.4 or 4.9 branch and features on the master
branch. Select the correct branch in the "base:" drop-down menu above.

Replace this notice with a short README for your feature/bugfix. This will help
people to understand your PR and can be used as a start for the documentation.
-->

This PR allows to customize page title and description same as it's allowed for news (#161) and events (c1970b099d6dc071811e3519f80e5c0400d8f6a2)

dmolineus committed:
Allow to override page title and description

dmolineus committed:
Implement custom robots support

Initially implemented in #1778 but transferred to this pr as requested in
https://github.com/contao/contao/pull/1778#pullrequestreview-437324426

leofeyer commented on faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php:
```diff
@@ -106,12 +106,29 @@ protected function compile()
-		if ($objFaq->question != '')
+		if ($objFaq->pageTitle)
+		{
+			$objPage->pageTitle = strip_tags(StringUtil::stripInsertTags($objFaq->pageTitle));
```
```suggestion
			$objPage->pageTitle = $objFaq->pageTitle;
```

leofeyer commented on faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php:
```diff
@@ -106,12 +106,29 @@ protected function compile()
+
+		if ($objFaq->description)
+		{
+			$objPage->description = $this->prepareMetaDescription($objFaq->description);
```
```suggestion
			$objPage->description = $objFaq->description;
```

leofeyer commented on faq-bundle/src/Resources/contao/models/FaqModel.php:
```diff
@@ -69,6 +75,9 @@
  * @method static Collection|FaqModel[]|FaqModel|null findByAnswer($val, $opt=array())
+ * @method static Collection|FaqModel[]|FaqModel|null findByPageTitle($val, $opt=array())
+ * @method static Collection|FaqModel[]|FaqModel|null findByRobots($val, $opt=array())
+ * @method static Collection|FaqModel[]|FaqModel|null findByDescription($val, $opt=array())
```
Please also add the `countBy*()` methods.

dmolineus replied:
Added with 5872196.

rabauss commented:
To have it similar to news + events - it would be nice to have `serpPreview` too, wouldn't it?

dmolineus committed:
Do not strip insert tags

Commit suggestions from @leofeyer
https://github.com/contao/contao/pull/1779#discussion_r447039554

Co-authored-by: Leo Feyer <github@contao.org>

dmolineus committed:
Do not preprocess description

Adopt suggestions made by @leofeyer https://github.com/contao/contao/pull/1779#discussion_r447039877

Co-authored-by: Leo Feyer <github@contao.org>

dmolineus committed:
Annotate new `countBy*` methods

dmolineus committed:
Add the serp preview for faqs

For faqs without an existing `jumpTo` page the serp preview is hidden

dmolineus commented:
> To have it similar to news + events - it would be nice to have `serpPreview` too, wouldn't it?

Added with 00ecc0c .

dmolineus commented:
All suggestions adopted and also added serp preview.

dmolineus committed:
Revert logic for serp preview

Remove field instead of adding it so it stays full customizable by changing the palette and or the user permissions

dmolineus committed:
Fulfill changed coding standards

dmolineus committed:
Yet another code style fix

WTH does the local code style check does not detect all issues?

leofeyer commented on faq-bundle/src/Resources/contao/dca/tl_faq.php:
```diff
@@ -486,6 +523,18 @@ public function checkPermission()
 
+	public function checkSerpPreview()
+	{
+		$objFaqCategory = FaqCategoryModel::findByPk(CURRENT_ID);
```
Not sure if we can/should use the model here. @aschempp Did this not cause problems in some cases?

dmolineus replied:
Changed with 1e3daf6

leofeyer commented on faq-bundle/src/Resources/contao/dca/tl_faq.php:
```diff
@@ -516,6 +565,35 @@ public function generateAlias($varValue, DataContainer $dc)
+	{
+		/** @var FaqCategoryModel $objCategory */
+		$objCategory = $objFaq->getRelated('pid');
+		$jumpTo = (int) $objCategory->jumpTo;
```
What if `$objCategory` is null?

dmolineus replied:
This shouldn't never happen in the backend view. In the sense of defensive programming I adjusted it e8307f0639aa9ecdd2eec8cf051307686415c1b3

leofeyer commented on faq-bundle/src/Resources/contao/dca/tl_faq.php:
```diff
@@ -516,6 +565,35 @@ public function generateAlias($varValue, DataContainer $dc)
+		// A jumpTo page is not mandatory for FAQ categories (see #6226) but required for the FAQ list module
+		if ($jumpTo < 1)
+		{
+			throw new Exception("FAQ categories without redirect page cannot be used in an FAQ list");
```
```suggestion
			throw new Exception('FAQ categories without redirect page cannot be used in an FAQ list');
```

leofeyer commented on faq-bundle/src/Resources/contao/dca/tl_faq.php:
```diff
@@ -516,6 +565,35 @@ public function generateAlias($varValue, DataContainer $dc)
+			return sprintf(preg_replace('/%(?!s)/', '%%', $strSuffix), ($objFaq->alias ?: $objFaq->id));
+		}
+
+		throw new Exception("FAQ categories without existing redirect page cannot be used in an FAQ list");
```
```suggestion
		throw new Exception('FAQ categories without existing redirect page cannot be used in an FAQ list');
```

aschempp commented:
@dmolineus can you finish this PR for Contao 4.11? 🙃

dmolineus commented:
> @dmolineus can you finish this PR for Contao 4.11? upside_down_face

I will, this PR is on my todo list.

dmolineus committed:
Do not use the faq category model

dmolineus committed:
Use single quotes

Co-authored-by: Leo Feyer <github@contao.org>

dmolineus committed:
Use single quotes

Co-authored-by: Leo Feyer <github@contao.org>

dmolineus committed:
Handle the case where category does not exist

dmolineus commented:
A changed issues mentioned in the feedback and rebased on the current master.

aschempp commented on faq-bundle/src/Resources/contao/dca/tl_faq.php:
```diff
@@ -520,6 +569,41 @@ public function generateAlias($varValue, DataContainer $dc)
+		// A jumpTo page is not mandatory for FAQ categories (see #6226) but required for the FAQ list module
+		if ($jumpTo < 1)
+		{
+			throw new Exception('FAQ categories without redirect page cannot be used in an FAQ list');
```
should this message be translated (dunno if thats the same for news etc)?

dmolineus replied:
This message is only shown if any third party extension uses the `getSerpUrl` manually. `tl_faq::checkSerpPreview` will remove the `serpPreview` field for faq categories without a redirect page.

leofeyer commented on faq-bundle/src/Resources/contao/modules/ModuleFaq.php:
```diff
@@ -103,6 +103,11 @@ public function getSearchablePages($arrPages, $intRoot=0, $blnIsSitemap=false)
 				{
 					while ($objItems->next())
 					{
+						if ($blnIsSitemap && $objItems->robots === 'noindex,nofollow')
```
This change seems unrelated. Can you please elaborate?

dmolineus replied:
The description of this PR is misleading. It also contains the robots setting introduced for news and events with https://github.com/contao/contao/pull/1778, see your comment here https://github.com/contao/contao/pull/1778#pullrequestreview-437324426

leofeyer replied:
I see. 👍🏻

leofeyer committed:
Merge branch 'master' into feature/faq-metadata

leofeyer committed:
Use an absolute URL in the SERP preview

leofeyer reviewed (dismissed):
Looks good to me. 👍🏻

leofeyer commented on faq-bundle/src/Resources/contao/dca/tl_faq.php:
```diff
@@ -486,6 +523,25 @@ public function checkPermission()
+		if (!$objFaqCategory->jumpTo)
+		{
+			PaletteManipulator::create()
+				->removeField(array('pageTitle', 'robots', 'description', 'serpPreview'), 'meta_legend')
```
@dmolineus I have adjusted this to remove all meta fields instead of just the `serpPreview` field when the FAQ category does not have a jumpTo page. Does this make sense to you as well?

dmolineus replied:
LGTM

leofeyer commented:
Thank you @dmolineus.
